src/OperationProcessor.ts

```diff
@@ -157,6 +158,63 @@ interface OperationInfo {
   missingAncestor?: VersionId;
 }
 
+/**
+ * An abstraction of a *complete* store for operations, exposing methods to store and
+ * subsequently retrieve operations using OperationInfo. Internally relies on a
+ * cache to lookup recent and/or heavily accessed operations; on a cache miss relies on
+ * an expensive CAS lookup to reconstruct the operation.
+ */
+class OperationStore {
+
+  private readonly operationCache: Cache<VersionId, WriteOperation>;
+
+  // Size for the operation cache; TODO: set from a config file?
+  private readonly operationCacheSize = 100000;
+
+  public constructor (private readonly cas: Cas) {
+    this.operationCache = getCache(this.operationCacheSize);
+  }
+
+  /**
+   * Store an operation in the store
+   */
+  public store (opHash: OperationHash, operation: WriteOperation) {
+    this.operationCache.store(opHash, operation);
+  }
+
+  /**
+   * Lookup an operation from the store
+   */
+  public async lookup (opHash: OperationHash, opInfo: OperationInfo): Promise<WriteOperation> {
```

Comment:
Design: Since this is an abstraction layer, I think opInfo should not be passed in for a lookup, else it defeats the purpose of an abstraction; ie. the caller shouldn't need to pass in data needed in case of a miss-hit by the internal cache. This probably leads to the need to store all the opInfo within the OperationStore, and to avoid duplication, the opHashToInfo probably can be moved within the OperationStore, which is probably the right decision following this design.

Reply:
Very nice suggestion. I implemented the suggestion - now lookup does not require an opInfo parameter.

I made one minor tweak to the second part of your suggestion. The OperationInfo interface contains a lot of details specific to OperationProcessor (e.g., status) which the OperationStore need not know about. I split the OperationInfo struct into two - OperationInfo (required by OperationProcessor) and OperationAccessInfo(required by OperationStore). The OperationAccessInfo struct contains informatoin such as batchFilehash which is required to reconstruct the operation from the cas.